test: migrate receive flow e2e infrastructure to TypeScript with shared types

chrisleewilcox opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
This PR migrates the complete receive flow e2e infrastructure from JavaScript to TypeScript, introducing shared type definitions and improving type safety across the entire test suite.

## What Changed

### 🔄 **Page Objects Converted:**
- `e2e/pages/Receive/PaymentRequestQrBottomSheet.js` → `.ts`
- `e2e/pages/Receive/RequestPaymentModal.js` → `.ts` 
- `e2e/pages/Receive/RequestPaymentView.js` → `.ts`
- `e2e/pages/Receive/SendLinkView.js` → `.ts`

### 🎯 **Selector Files Converted:**
- `e2e/selectors/Receive/RequestPaymentModal.selectors.js` → `.ts`
- `e2e/selectors/Receive/RequestPaymentView.selectors.js` → `.ts`
- `e2e/selectors/Receive/SendLinkView.selectors.js` → `.ts`

### 🧪 **Test Spec Converted:**
- `e2e/specs/wallet/request-token-flow.spec.js` → `.ts`

### 🆕 **New Shared Types Infrastructure:**
- `e2e/types/detox.d.ts` - Centralized Detox element type definitions

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/MMQA-379

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**
![image](https://github.com/user-attachments/assets/4b1ddbe7-626a-40eb-8d8c-fdd12c764107)

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

chrisleewilcox committed:
feat: add PaymentRequestQrBottomSheet for QR modal interactions

This commit introduces a new class, PaymentRequestQrBottomSheet, which encapsulates the functionality for interacting with the QR modal in the payment request flow. It includes methods to access the modal container and close button, as well as a method to tap the close button using defined gestures.

chrisleewilcox committed:
feat: add RequestPaymentModal class for payment request interactions

This commit introduces a new class, RequestPaymentModal, which encapsulates the functionality for interacting with the request payment modal. It includes methods to access the request payment button and tap it using defined gestures, enhancing the end-to-end testing capabilities for payment requests.

chrisleewilcox committed:
feat: add RequestPaymentView class for payment request interactions

This commit introduces a new class, RequestPaymentView, which encapsulates the functionality for interacting with the request payment view. It includes methods to access various UI elements and perform actions such as tapping buttons and entering text, enhancing the end-to-end testing capabilities for payment requests.

chrisleewilcox committed:
feat: add SendLinkView class for link sending interactions

This commit introduces a new class, SendLinkView, which encapsulates the functionality for interacting with the send link view. It includes methods to access various UI elements such as the QR modal and buttons, enhancing the end-to-end testing capabilities for link sending interactions.

chrisleewilcox committed:
feat: add selectors for RequestPaymentModal, RequestPaymentView, and SendLinkView

This commit introduces new selector files for the RequestPaymentModal, RequestPaymentView, and SendLinkView components. These selectors define constants for various UI elements, enhancing the end-to-end testing capabilities for payment request and link sending interactions.

cortisiko commented on e2e/pages/Receive/RequestPaymentModal.ts:
```diff
@@ -2,18 +2,20 @@ import { RequestPaymentModalSelectorsIDs } from '../../selectors/Receive/Request
 import Matchers from '../../utils/Matchers';
 import Gestures from '../../utils/Gestures';
 
+type DetoxElement = Promise<Detox.IndexableNativeElement | Detox.NativeElement | Detox.IndexableSystemElement>;
```
I see this is used in multiple files. I take it in future files we will be using these type aliases more. How about creating a `detox.types.ts` that exports these?

cortisiko commented on e2e/pages/Receive/RequestPaymentView.ts:
```diff
@@ -2,51 +2,55 @@ import { RequestPaymentViewSelectors } from '../../selectors/Receive/RequestPaym
 import Gestures from '../../utils/Gestures';
 
+type DetoxElement = Promise<Detox.IndexableNativeElement | Detox.NativeElement | Detox.IndexableSystemElement>;
+type TappableElement = Promise<Detox.IndexableNativeElement | Detox.SystemElement>;
```
I see this is used in multiple files. I take it in future files we will be using these type aliases more. How about creating a detox.types.ts that exports these?

chrisleewilcox replied:
Added ./e2e/types folder with detox.d.ts.

cortisiko commented on e2e/pages/Receive/RequestPaymentView.ts:
```diff
@@ -2,51 +2,55 @@ import { RequestPaymentViewSelectors } from '../../selectors/Receive/RequestPaym
 
+type DetoxElement = Promise<Detox.IndexableNativeElement | Detox.NativeElement | Detox.IndexableSystemElement>;
+type TappableElement = Promise<Detox.IndexableNativeElement | Detox.SystemElement>;
+type TypableElement = Promise<Detox.IndexableNativeElement>;
```
I see this is used in multiple files. I take it in future files we will be using these type aliases more. How about creating a detox.types.ts that exports these?

cortisiko reviewed (commented):
small comment. Keep it going with the .ts PRs! 

🚗 💨

chrisleewilcox committed:
test: convert request-token-flow.spec.js to TypeScript

Add proper type annotations for async functions, constants, and Jest lifecycle methods. Maintains full test functionality with enhanced type safety.

chrisleewilcox committed:
refactor: centralize Detox element types for better maintainability

This commit introduces a new file for common Detox element types, allowing for cleaner imports across multiple files. The previous inline type definitions in PaymentRequestQrBottomSheet, RequestPaymentModal, RequestPaymentView, and SendLinkView have been replaced with imports from the new types file, enhancing code readability and maintainability.

chrisleewilcox committed:
fix: add TypeScript workaround for FixtureBuilder user state in request-token-flow.spec.ts

This commit introduces a TypeScript workaround in the request-token-flow.spec.ts file to address the lack of exposed user state types in FixtureBuilder. The workaround sets the seedphraseBackedUp property to false, ensuring proper test setup and execution.

cortisiko reviewed (dismissed):
🏇 💨

chrisleewilcox committed:
refactor: remove inline Detox element type imports for improved clarity

This commit removes the inline imports of Detox element types from PaymentRequestQrBottomSheet, RequestPaymentModal, RequestPaymentView, and SendLinkView. The types are now globally available, simplifying the code and enhancing maintainability across the e2e test files.

chrisleewilcox committed:
refactor: enhance global Detox types for e2e tests

This commit expands the global Detox types in detox.d.ts to include common, individual, configuration, and matcher types, improving type clarity and usability across e2e test files. This change simplifies type management and enhances maintainability in the testing framework.

chrisleewilcox committed:
refactor: remove unused Detox type import in NotificationMenuView

This commit removes the unused import of the IndexableNativeElement type in the NotificationMenuView file, streamlining the code and improving clarity. This change aligns with the ongoing effort to enhance maintainability across e2e test files.

chrisleewilcox committed:
refactor: update Detox element type usage in bridge-functionality.spec.ts

This commit refactors the bridge-functionality.spec.ts file to replace inline Detox element type imports with the globally available IndexableNativeElement type. This change improves code clarity and maintainability, aligning with recent enhancements to type management across e2e test files.

chrisleewilcox committed:
refactor: remove unused Detox type import in enable-notifications-after-onboarding.spec.ts

This commit removes the unused import of the DeviceLaunchAppConfig type in the enable-notifications-after-onboarding.spec.ts file, streamlining the code and improving clarity. This change continues the effort to enhance maintainability across e2e test files.
